clientlibrary/checkpoint/dynamodb-checkpointer_test.go

```diff
@@ -156,10 +157,135 @@ func TestGetLeaseAquired(t *testing.T) {
 	assert.Equal(t, "", status.GetLeaseOwner())
 }
 
+func TestGetLeaseShardClaimed(t *testing.T) {
+	svc := &mockDynamoDB{
+		tableExist: true,
+		item: map[string]*dynamodb.AttributeValue{
+			CLAIM_REQUEST_KEY: {S: aws.String("ijkl-mnop")},
+		},
+	}
+	kclConfig := cfg.NewKinesisClientLibConfig("appName", "test", "us-west-2", "abc").
+		WithInitialPositionInStream(cfg.LATEST).
+		WithMaxRecords(10).
+		WithMaxLeasesForWorker(1).
+		WithShardSyncIntervalMillis(5000).
+		WithFailoverTimeMillis(300000).
+		WithLeaseStealing(true)
+
+	checkpoint := NewDynamoCheckpoint(kclConfig).WithDynamoDB(svc)
+	checkpoint.Init()
+	err := checkpoint.GetLease(&par.ShardStatus{
+		ID:         "0001",
+		Checkpoint: "",
+		Mux:        &sync.Mutex{},
+	}, "abcd-efgh")
+	if err == nil || err.Error() != ErrShardClaimed {
+		t.Errorf("Got a lease when it was already claimed by by ijkl-mnop: %s", err)
+	}
+
+	err = checkpoint.GetLease(&par.ShardStatus{
+		ID:         "0001",
+		Checkpoint: "",
+		Mux:        &sync.Mutex{},
+	}, "ijkl-mnop")
+	if err != nil {
+		t.Errorf("Error getting lease %s", err)
+	}
+}
+
+func TestFetchCheckpointWithStealing(t *testing.T) {
+	future := time.Now().AddDate(0, 1, 0)
+
+	svc := &mockDynamoDB{
+		tableExist: true,
+		item: map[string]*dynamodb.AttributeValue{
+			CHECKPOINT_SEQUENCE_NUMBER_KEY: {S: aws.String("deadbeef")},
+			LEASE_OWNER_KEY:                {S: aws.String("abcd-efgh")},
+			LEASE_TIMEOUT_KEY: {
+				S: aws.String(future.Format(time.RFC3339)),
+			},
+		},
+	}
+
+	kclConfig := cfg.NewKinesisClientLibConfig("appName", "test", "us-west-2", "abc").
+		WithInitialPositionInStream(cfg.LATEST).
+		WithMaxRecords(10).
+		WithMaxLeasesForWorker(1).
+		WithShardSyncIntervalMillis(5000).
+		WithFailoverTimeMillis(300000).
+		WithLeaseStealing(true)
+
+	checkpoint := NewDynamoCheckpoint(kclConfig).WithDynamoDB(svc)
+	checkpoint.Init()
+
+	status := &par.ShardStatus{
+		ID:           "0001",
+		Checkpoint:   "",
+		LeaseTimeout: time.Now(),
+		Mux:          &sync.Mutex{},
+	}
+
+	checkpoint.FetchCheckpoint(status)
+
+	leaseTimeout, _ := time.Parse(time.RFC3339, *svc.item[LEASE_TIMEOUT_KEY].S)
+	assert.Equal(t, leaseTimeout, status.LeaseTimeout)
+}
+
+func TestGetLeaseConditional(t *testing.T) {
+	svc := &mockDynamoDB{tableExist: true, item: map[string]*dynamodb.AttributeValue{}}
+	kclConfig := cfg.NewKinesisClientLibConfig("appName", "test", "us-west-2", "abc").
+		WithInitialPositionInStream(cfg.LATEST).
+		WithMaxRecords(10).
+		WithMaxLeasesForWorker(1).
```

Comment:
Test should be done with a larger number of shards (10+) and each worker need to have larger number of shard. Or, just don't set it at all to have max default value.

Reply:
Not really sure what you mean here. This is just a unit test of the the leasestealer. There is no actual instance of a worker in the test at all. We can remove the `WithMaxLeasesForWorker(1)`, but I'm not sure what you mean about "Test should be done with a larger number of shards (10+)` since there are no actual shards."

Reply:
Create a stream on AWS with 10 shards.

Reply:
Why? There is no worker. This is a unit test.